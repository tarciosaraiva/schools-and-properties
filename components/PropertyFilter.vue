<template>
  <modal @close="closeFn">
    <template #header>
      <h2>Property filter</h2>
    </template>
    <template #body>
      <form>
        <div class="pl-row">
          <div class="pl-col">
            <p>Property type</p>
          </div>
        </div>
        <div class="pl-row">
          <div class="pl-col-md-4">
            <label for="p_house_type">
            <input
              id="p_house_type"
              v-model="pTypes"
              type="checkbox"
              value="House"
            /> House</label>
          </div>
          <div class="pl-col-md-4">
            <label for="p_townhouse_type">
            <input
              id="p_townhouse_type"
              v-model="pTypes"
              type="checkbox"
              value="Townhouse"
            /> Townhouse </label>
          </div>
          <div class="pl-col-md-4">
            <label for="p_unit_type">
            <input
              id="p_unit_type"
              v-model="pTypes"
              type="checkbox"
              value="ApartmentUnitFlat"
            /> Unit </label>
          </div>
        </div>
        <div class="pl-row">
          <div class="pl-col-md-6">
            <label for="p_rooms">Rooms</label>
            <select id="p_rooms" v-model="pRooms">
              <option value="1">1+</option>
              <option value="2">2+</option>
              <option value="3">3+</option>
              <option value="4">4+</option>
            </select>
          </div>
          <div class="pl-col-md-6">
            <label for="p_bathrooms">Bathrooms</label>
            <select id="p_bathrooms" v-model="pBathrooms">
              <option value="1">1+</option>
              <option value="2">2+</option>
              <option value="3">3+</option>
              <option value="4">4+</option>
            </select>
          </div>
        </div>
        <div class="pl-row">
          <div class="pl-col-md-6">
            <label for="p_car_spaces">Car spaces</label>
            <select id="p_car_spaces" v-model="pCarSpaces">
              <option value="1">1+</option>
              <option value="2">2+</option>
              <option value="3">3+</option>
              <option value="4">4+</option>
            </select>
          </div>
          <div class="pl-col-md-6">
            <label for="p_max_price">Max price</label>
            <select id="p_max_price" v-model="pMaxPrice">
              <option value="500000">500 K</option>
              <option value="600000">600 K</option>
              <option value="700000">700 K</option>
              <option value="800000">800 K</option>
              <option value="900000">900 K</option>
              <option value="1000000">1 M</option>
              <option value="1100000">1.1 M</option>
              <option value="1200000">1.2 M</option>
              <option value="1300000">1.3 M</option>
              <option value="1400000">1.4 M</option>
              <option value="1500000">1.5 M</option>
            </select>
          </div>
        </div>
      </form>
    </template>
  </modal>
</template>

<script lang="ts">
import Vue from 'vue'
import { mapGetters } from 'vuex'

import Modal from '~/components/Modal.vue'

export default Vue.extend({
  components: { Modal },
  props: {
    closeFn: Function
  },
  computed: {
    ...mapGetters(['propertiesFilter']),
    pTypes: {
      get() {
        return this.propertiesFilter.types
      },
      set(value) {
        this.$store.commit('UPDATE_PROPERTY_FILTER_TYPE', value)
      },
    },
    pRooms: {
      get() {
        return this.propertiesFilter.rooms
      },
      set(value) {
        this.$store.commit('UPDATE_PROPERTY_FILTER_ROOMS', value)
      },
    },
    pBathrooms: {
      get() {
        return this.propertiesFilter.bathrooms
      },
      set(value) {
        this.$store.commit('UPDATE_PROPERTY_FILTER_BATHROOMS', value)
      },
    },
    pCarSpaces: {
      get() {
        return this.propertiesFilter.carSpaces
      },
      set(value) {
        this.$store.commit('UPDATE_PROPERTY_FILTER_CAR_SPACES', value)
      },
    },
    pMaxPrice: {
      get() {
        return this.propertiesFilter.maxPrice
      },
      set(value) {
        this.$store.commit('UPDATE_PROPERTY_FILTER_MAX_PRICE', value)
      },
    }
  },
})
</script>
