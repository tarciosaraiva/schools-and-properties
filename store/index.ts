import { GetterTree, ActionTree, MutationTree } from 'vuex'

interface ApplicationFilter {
  properties: PropertiesFilter,
  schools: SchoolsFilter
}

interface PropertiesFilter {
  types: string[],
  rooms: number,
  bathrooms: number,
  carSpaces: number,
  maxPrice: number
}

export interface SchoolsFilter {
  educationSector: string,
  zone: string,
  primary: SchoolLevelRating,
  secondary: SchoolLevelRating,
}

interface SchoolLevelRating {
  plot: boolean,
  rating: number,
  englishRating: number,
  mathsRating: number
}

export interface PropertyListing {
  id: number,
  price: string,
  area: string
  bathrooms: number,
  bedrooms: number,
  carspaces: number,
  displayableAddress: string,
  features: string[],
  landArea: number,
  latitude: number,
  longitude: number,
  postcode: string,
  propertyType: string
  region: string,
  state: string,
  street: string,
  streetNumber: string,
  suburb: string,
  unitNumber: string,
  media: string[],
  listingSlug: string
}

export interface School {
  schoolName: string,
  educationSector: string,
  schoolType: string,
  phoneNumber: string,
  primaryOverallScore: string,
  primaryEnglishScore: string,
  primaryMathsScore: string,
  primaryEnrolments: string,
  secondaryOverallScore: string,
  secondaryEnglishScore: string,
  secondaryMathsScore: string,
  secondaryEnrolments: string
}

export const state = () => ({
  bounds: {
    ne: { lat: 0, lon: 0 },
    nw: { lat: 0, lon: 0 },
    se: { lat: 0, lon: 0 },
    sw: { lat: 0, lon: 0 },
  },
  geocodeSuggestions: [] as any,
  list: [] as PropertyListing[],
  filter: {
    properties: {
      types: ['House', 'Townhouse'],
      rooms: 3,
      bathrooms: 2,
      carSpaces: 1,
      maxPrice: 900000
    },
    schools: {
      educationSector: 'all',
      zone: 'none',
      primary: {
        plot: true,
        rating: 95,
        englishRating: 4,
        mathsRating: 4
      },
      secondary: {
        plot: true,
        rating: 95,
        englishRating: 4,
        mathsRating: 4
      }
    }
  } as ApplicationFilter
})

export type RootState = ReturnType<typeof state>

export const getters: GetterTree<RootState, RootState> = {
  listings: state => state.list,
  schoolsFilter: state => state.filter.schools,
  propertiesFilter: state => state.filter.properties,
  locations: state => state.geocodeSuggestions,
  listingPois: state => state.list.map(l => ({
    type: 'Feature',
    id: l.id,
    geometry: {
      type: 'Point',
      coordinates: [l.longitude, l.latitude]
    },
    properties: { ...l }
  }))
}

export const mutations: MutationTree<RootState> = {
  ADD_LISTINGS (state, listings: any[]) {
    const propertyListings = listings
      .filter(l => l.type === 'PropertyListing')
      .map(({ listing }) => {
        const propertyListing: PropertyListing = {
          id: listing.id,
          price: listing.priceDetails?.displayPrice,
          area: listing.propertyDetails.area,
          bathrooms: listing.propertyDetails.bathrooms,
          bedrooms: listing.propertyDetails.bedrooms,
          carspaces: listing.propertyDetails.carspaces,
          displayableAddress: listing.propertyDetails.displayableAddress,
          features: listing.propertyDetails.features,
          landArea: listing.propertyDetails.landArea,
          latitude: listing.propertyDetails.latitude,
          longitude: listing.propertyDetails.longitude,
          postcode: listing.propertyDetails.postcode,
          propertyType: listing.propertyDetails.propertyType,
          region: listing.propertyDetails.region,
          state: listing.propertyDetails.state,
          street: listing.propertyDetails.street,
          streetNumber: listing.propertyDetails.streetNumber,
          suburb: listing.propertyDetails.suburb,
          unitNumber: listing.propertyDetails.unitNumber,
          media: listing.media.map((m: any) => m.url),
          listingSlug: listing.listingSlug
        }

        return propertyListing
      })

    state.list = propertyListings
  },
  UPDATE_PROPERTY_FILTER_TYPE (state, payload) {
    state.filter.properties.types = payload
  },
  UPDATE_PROPERTY_FILTER_ROOMS (state, payload) {
    state.filter.properties.rooms = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_BATHROOMS (state, payload) {
    state.filter.properties.bathrooms = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_CAR_SPACES (state, payload) {
    state.filter.properties.carSpaces = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_MAX_PRICE (state, payload) {
    state.filter.properties.maxPrice = Number(payload)
  },
  UPDATE_BOUNDING_BOX (state, payload) {
    state.bounds.nw.lat = payload.getNorthWest().lat
    state.bounds.nw.lon = payload.getNorthWest().lng
    state.bounds.ne.lat = payload.getNorthEast().lat
    state.bounds.ne.lon = payload.getNorthEast().lng
    state.bounds.se.lat = payload.getSouthEast().lat
    state.bounds.se.lon = payload.getSouthEast().lng
    state.bounds.sw.lat = payload.getSouthWest().lat
    state.bounds.sw.lon = payload.getSouthWest().lng
  },
  UPDATE_SCHOOL_EDUCATION_SECTOR (state, payload) {
    state.filter.schools.educationSector = payload
  },
  UPDATE_SCHOOL_ZONE (state, payload) {
    state.filter.schools.zone = payload
  },
  UPDATE_PLOT_PRIMARY_SCHOOLS (state, payload) {
    state.filter.schools.primary.plot = payload
  },
  UPDATE_SCHOOL_PRIMARY_RATING (state, payload) {
    state.filter.schools.primary.rating = Number(payload)
  },
  UPDATE_SCHOOL_PRIMARY_ENG_RATING (state, payload) {
    state.filter.schools.primary.englishRating = Number(payload)
  },
  UPDATE_SCHOOL_PRIMARY_MATH_RATING (state, payload) {
    state.filter.schools.primary.mathsRating = Number(payload)
  },
  UPDATE_PLOT_SECONDARY_SCHOOLS (state, payload) {
    state.filter.schools.secondary.plot = payload
  },
  UPDATE_SCHOOL_SECONDARY_RATING (state, payload) {
    state.filter.schools.secondary.rating = Number(payload)
  },
  UPDATE_SCHOOL_SECONDARY_ENG_RATING (state, payload) {
    state.filter.schools.secondary.englishRating = Number(payload)
  },
  UPDATE_SCHOOL_SECONDARY_MATH_RATING (state, payload) {
    state.filter.schools.secondary.mathsRating = Number(payload)
  },
  ADD_GEOCODE_RESULTS (state, payload) {
    state.geocodeSuggestions = payload.map((f: any) => ({
      center: f.center,
      name: f.place_name,
      placeType: f.place_type
    }))
  }
}

export const actions: ActionTree<RootState, RootState> = {
  async loadListings ({ commit, state }) {
    const { filter, bounds } = state

    const payload = {
      listingType: 'Sale',
      pageSize: 200,
      propertyTypes: filter.properties.types,
      minBedrooms: filter.properties.rooms,
      minBathrooms: filter.properties.bathrooms,
      minCarspaces: filter.properties.carSpaces,
      maxPrice: filter.properties.maxPrice,
      locations: [{ state: 'VIC' }],
      geoWindow: { box: { topLeft: bounds.nw, bottomRight: bounds.se } }
    }

    const listings = await this.$axios.$post('https://api.domain.com.au/v1/listings/residential/_search', payload, { headers: {'X-Api-Key': process.env.domainApiKey } })
    commit('ADD_LISTINGS', listings)
  },

  async geocode ({ commit }, searchTerm) {
    if (searchTerm) {
      const params = {
        key: process.env.mapTilerSecret,
        bbox: '141.24847451171752,-39.624732230379166,148.68072548828115,-34.32302549095252',
        language: 'en'
      }

      const geocodeResults = await this.$axios.$get(`https://api.maptiler.com/geocoding/${searchTerm}.json`, { params })
      const schoolsResults = await this.$axios.$get(`//${process.env.domain}/schools/schools.json`)
      const mappedSchoolResults = schoolsResults.map((s: any) => ({
        center: [s.lon, s.lat],
        place_name: s.schoolName,
        place_type: 'school'
      })).filter((s: any) => s.place_name.toLowerCase().includes(searchTerm))

      commit('ADD_GEOCODE_RESULTS', [...geocodeResults.features, ...mappedSchoolResults])
    } else {
      commit('ADD_GEOCODE_RESULTS', [])
    }
  },

  setBoundingBox({ commit }, payload) {
    commit('UPDATE_BOUNDING_BOX', payload)
  }
}
