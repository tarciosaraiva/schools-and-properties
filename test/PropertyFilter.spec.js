import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'

import { state, getters } from '@/store'
import PropertyFilter from '~/components/PropertyFilter.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('PropertyFilter', () => {
  let store
  let wrapper
  let propsData
  let mutations

  beforeEach(() => {
    propsData = { closeFn: jest.fn() }
    mutations = {
      'UPDATE_PROPERTY_FILTER_TYPE': jest.fn(),
      'UPDATE_PROPERTY_FILTER_ROOMS': jest.fn(),
      'UPDATE_PROPERTY_FILTER_BATHROOMS': jest.fn(),
      'UPDATE_PROPERTY_FILTER_CAR_SPACES': jest.fn(),
      'UPDATE_PROPERTY_FILTER_MAX_PRICE': jest.fn(),
    }
    store = new Vuex.Store({ state, mutations, getters })
    wrapper = mount(PropertyFilter, { propsData, store, localVue })
  })

  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders correctly with props and store', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('property types default values matches state', () => {
    const houseInput = wrapper.find('#p_house_type')
    expect(houseInput.element.checked).toBeTruthy()

    const townhouseInput = wrapper.find('#p_townhouse_type')
    expect(townhouseInput.element.checked).toBeTruthy()

    const unitInput = wrapper.find('#p_unit_type')
    expect(unitInput.element.checked).toBeFalsy()
  })

  test('rooms default value match state', () => {
    const input = wrapper.find('#p_rooms')
    expect(Number(input.element.value)).toBe(store.state.filter.properties.rooms)
  })

  test('bathrooms default value match state', () => {
    const input = wrapper.find('#p_bathrooms')
    expect(Number(input.element.value)).toBe(store.state.filter.properties.bathrooms)
  })

  test('car spaces default value match state', () => {
    const input = wrapper.find('#p_car_spaces')
    expect(Number(input.element.value)).toBe(store.state.filter.properties.carSpaces)
  })

  test('max price default value match state', () => {
    const input = wrapper.find('#p_max_price')
    expect(Number(input.element.value)).toBe(store.state.filter.properties.maxPrice)
  })

  test('should have committed mutation "UPDATE_PROPERTY_FILTER_TYPE" to the store when rating changes', async () => {
    const input = wrapper.find('#p_townhouse_type')
    await input.setChecked(false)
    expect(mutations.UPDATE_PROPERTY_FILTER_TYPE).toHaveBeenCalledWith(store.state, ['House'])
  })

  test.each([
    ['UPDATE_PROPERTY_FILTER_ROOMS', '#p_rooms', '2'],
    ['UPDATE_PROPERTY_FILTER_BATHROOMS', '#p_bathrooms', '3'],
    ['UPDATE_PROPERTY_FILTER_CAR_SPACES', '#p_car_spaces', '2'],
    ['UPDATE_PROPERTY_FILTER_MAX_PRICE', '#p_max_price', '500000'],
  ])(
    'should have committed mutation %p to the store when rating changes',
    (mutation, elementId, value) => {
      const input = wrapper.find(elementId)
      input.element.value = value
      input.trigger('change')

      expect(mutations[mutation]).toHaveBeenCalledWith(store.state, value)
    }
  )
})
