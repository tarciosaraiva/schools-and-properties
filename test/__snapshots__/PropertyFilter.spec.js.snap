// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyFilter renders correctly with props and store 1`] = `
<transition-stub name="modal">
  <div class="modal-mask">
    <div class="modal-wrapper">
      <div class="modal-container">
        <div class="modal-header">
          <h2>Property filter</h2>
        </div>
        <div class="modal-body">
          <form>
            <div class="pl-row">
              <div class="pl-col">
                <p>Property type</p>
              </div>
            </div>
            <div class="pl-row">
              <div class="pl-col-md-4">
                <label for="p_house_type">
                  <input id="p_house_type" type="checkbox" value="House"> House
                </label>
              </div>
              <div class="pl-col-md-4">
                <label for="p_townhouse_type">
                  <input id="p_townhouse_type" type="checkbox" value="Townhouse"> Townhouse
                </label>
              </div>
              <div class="pl-col-md-4">
                <label for="p_unit_type">
                  <input id="p_unit_type" type="checkbox" value="ApartmentUnitFlat"> Unit
                </label>
              </div>
            </div>
            <div class="pl-row">
              <div class="pl-col-md-6">
                <label for="p_rooms">Rooms</label>
                <select id="p_rooms">
                  <option value="1">1+</option>
                  <option value="2">2+</option>
                  <option value="3">3+</option>
                  <option value="4">4+</option>
                </select>
              </div>
              <div class="pl-col-md-6">
                <label for="p_bathrooms">Bathrooms</label>
                <select id="p_bathrooms">
                  <option value="1">1+</option>
                  <option value="2">2+</option>
                  <option value="3">3+</option>
                  <option value="4">4+</option>
                </select>
              </div>
            </div>
            <div class="pl-row">
              <div class="pl-col-md-6">
                <label for="p_car_spaces">Car spaces</label>
                <select id="p_car_spaces">
                  <option value="1">1+</option>
                  <option value="2">2+</option>
                  <option value="3">3+</option>
                  <option value="4">4+</option>
                </select>
              </div>
              <div class="pl-col-md-6">
                <label for="p_max_price">Max price</label>
                <select id="p_max_price">
                  <option value="500000">500 K</option>
                  <option value="600000">600 K</option>
                  <option value="700000">700 K</option>
                  <option value="800000">800 K</option>
                  <option value="900000">900 K</option>
                  <option value="1000000">1 M</option>
                  <option value="1100000">1.1 M</option>
                  <option value="1200000">1.2 M</option>
                  <option value="1300000">1.3 M</option>
                  <option value="1400000">1.4 M</option>
                  <option value="1500000">1.5 M</option>
                </select>
              </div>
            </div>
          </form>
        </div>
        <div class="modal-footer text-right">
          <button class="pl-button full small">
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</transition-stub>
`;
